### PR TITLE
Task metadata functional test

### PR DESCRIPTION
**Pull request by aaithal**
### Summary

Adds a functional test for task metadata endpoint

### Implementation details

Added a sample app for testing task metadata endpoint

### Testing

- [x] Builds on Linux (`make release`)
- [x] Builds on Windows (`go build -out amazon-ecs-agent.exe ./agent`)
- [x] Unit tests on Linux (`make test`) pass
- [x] Unit tests on Windows (`go test -timeout=25s ./agent/...`) pass
- [x] Integration tests on Linux (`make run-integ-tests`) pass
- [x] Integration tests on Windows (`.\scripts\run-integ-tests.ps1`) pass
- [x] Functional tests on Linux (`make run-functional-tests`) pass
- [x] Functional tests on Windows (`.\scripts\run-functional-tests.ps1`) pass

New tests cover the changes: Yes
### Licensing

This contribution is under the terms of the Apache 2.0 License: Yes

### CODE REVIEW

File: agent/functional_tests/tests/functionaltests_unix_test.go [hunk 1]
```diff
@@ -669,3 +669,24 @@ func TestMetadataServiceValidator(t *testing.T) {
 
 	assert.Equal(t, 42, exitCode, fmt.Sprintf("Expected exit code of 42; got %d", exitCode))
 }
+
+func TestTaskMetadataValidator(t *testing.T) {
+	RequireDockerVersion(t, ">=17.06.0-ce")
```
Comment (richardpen) [inline agent/functional_tests/tests/functionaltests_unix_test.go]: Can you add a comment here, why this requires docker version `>=17.06.0-ce`? also I'm not sure if `RequireDockerVersion` is smart enough to handle `-ce`, I think you can just `>=17.06.0`.

File: agent/functional_tests/tests/functionaltests_unix_test.go [hunk 2]
```diff
@@ -669,3 +669,24 @@ func TestMetadataServiceValidator(t *testing.T) {
...
+	agent := RunAgent(t, &AgentOptions{EnableTaskENI: true})
+	defer agent.Cleanup()
+
+	task, err := agent.StartAWSVPCTask("taskmetadata-validator-awsvpc")
+	require.NoError(t, err, "Unable to start task with 'awsvpc' network mode")
+	defer func() {
+		if err := task.Stop(); err != nil {
+			return
+		}
+		task.WaitStopped(2 * time.Minute)
```
Comment (richardpen) [inline agent/functional_tests/tests/functionaltests_unix_test.go]: Can you use the constant `waitTaskStateChangeDuration` instead.

File: agent/functional_tests/tests/functionaltests_unix_test.go [hunk 3]
```diff
@@ -669,3 +669,24 @@ func TestMetadataServiceValidator(t *testing.T) {
...
+	}()
+
+	err = task.WaitStopped(3 * time.Minute)
```
Comment (richardpen) [inline agent/functional_tests/tests/functionaltests_unix_test.go]: same as above.

File: misc/taskmetadata-validator/Makefile [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+#
+# or in the "license" file accompanying this file. This file is distributed
+# on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+# express or implied. See the License for the specific language governing
+# permissions and limitations under the License.
+.PHONY: all clean docker image
+
+all: taskmetadata-validator image
+
+#VENDOR_DIR=$(shell echo $(cd ../../agent/vendor; pwd))
```
Comment (richardpen) [inline misc/taskmetadata-validator/Makefile]: this comment needs to be deleted. also where is `VENDOR_DIR` used?

File: misc/taskmetadata-validator/build-in-docker [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+AGENT_VENDOR_DIR=$(echo $(cd ${CUR_DIR}/../../agent/vendor;pwd))
+
+docker run \
+	-v ${CUR_DIR}:/out \
+	-v ${CUR_DIR}/taskmetadata-validator.go:/in/taskmetadata-validator.go \
+	-v ${AGENT_VENDOR_DIR}:/gopath/src \
+	-e CGO_ENABLED=0 \
+	-e GOPATH=/go:/gopath \
+	golang:1.7 go build -a -x -ldflags '-s' -o /out/taskmetadata-validator /in/taskmetadata-validator.go
+
```
Comment (richardpen) [inline misc/taskmetadata-validator/build-in-docker]: blank line can be removed, also what's the `'-s'` here?

File: misc/taskmetadata-validator/Makefile [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+VENDOR_DIR=$(shell echo $(cd ../../agent/vendor; pwd))
+taskmetadata-validator: taskmetadata-validator.go
```
Comment (richardpen) [inline misc/taskmetadata-validator/Makefile]: I'm not an makefile expert, is it normal to put a file as dependency?

File: misc/taskmetadata-validator/Makefile [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+# not use this file except in compliance with the License. A copy of the
+# License is located at
+#
+#	http://aws.amazon.com/apache2.0/
+#
+# or in the "license" file accompanying this file. This file is distributed
+# on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+# express or implied. See the License for the specific language governing
+# permissions and limitations under the License.
+.PHONY: all clean docker image
```
Comment (richardpen) [inline misc/taskmetadata-validator/Makefile]: why is `docker` a phony here?

File: agent/functional_tests/tests/functionaltests_unix_test.go (hunk 1 above)
Comment (aaithal) [reply agent/functional_tests/tests/functionaltests_unix_test.go]: agent needs to be started with `--init`, which is required for 'awsvpc' networking mode

File: misc/taskmetadata-validator/Makefile (hunk 7 above)
Comment (aaithal) [reply misc/taskmetadata-validator/Makefile]: all non source file targets are `phony` target. please read makefile documentation for more details

File: misc/taskmetadata-validator/Makefile (hunk 6 above)
Comment (aaithal) [reply misc/taskmetadata-validator/Makefile]: it's the whole premise of makefiles :)

File: misc/taskmetadata-validator/build-in-docker (hunk 5 above)
Comment (aaithal) [reply misc/taskmetadata-validator/build-in-docker]: denotes a static binary is being built

**Pull request merged**